### PR TITLE
chore: refactor auto scaling resources

### PR DESCRIPTION
**Pull request by ShiChangkuo**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

* impl resources with Context
* using validation func in plugin sdk
* update acc testing names

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
$ make testacc TEST='./huaweicloud/services/acceptance/as' TESTARGS='-run TestAccAS'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccAS -timeout 360m -parallel 4
=== RUN   TestAccASConfiguration_basic
=== PAUSE TestAccASConfiguration_basic
=== RUN   TestAccASGroup_basic
=== PAUSE TestAccASGroup_basic
=== RUN   TestAccASGroup_withEpsId
=== PAUSE TestAccASGroup_withEpsId
=== RUN   TestAccASLifecycleHook_basic
=== PAUSE TestAccASLifecycleHook_basic
=== RUN   TestAccASPolicy_basic
=== PAUSE TestAccASPolicy_basic
=== CONT  TestAccASConfiguration_basic
=== CONT  TestAccASLifecycleHook_basic
=== CONT  TestAccASGroup_withEpsId
    acceptance.go:347: This environment does not support Enterprise Project ID tests
=== CONT  TestAccASGroup_basic
--- SKIP: TestAccASGroup_withEpsId (0.00s)
=== CONT  TestAccASPolicy_basic
--- PASS: TestAccASConfiguration_basic (43.37s)
--- PASS: TestAccASPolicy_basic (57.94s)
--- PASS: TestAccASLifecycleHook_basic (200.23s)
--- PASS: TestAccASGroup_basic (239.09s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        239.163s
```

### CODE REVIEW

File: huaweicloud/services/as/resource_huaweicloud_as_configuration.go [hunk 1]
```diff
@@ -280,18 +290,17 @@ func getInstanceConfig(configDataMap map[string]interface{}) (configurations.Ins
 	return instanceConfigOpts, nil
 }
 
-func resourceASConfigurationCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceASConfigurationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	asClient, err := config.AutoscalingV1Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
+		return diag.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_configuration.go]: pls  remove huawei

File: huaweicloud/services/as/resource_huaweicloud_as_configuration.go [hunk 2]
```diff
@@ -301,51 +310,52 @@ func resourceASConfigurationCreate(d *schema.ResourceData, meta interface{}) err
...
+	return resourceASConfigurationRead(ctx, d, meta)
 }
 
-func resourceASConfigurationRead(d *schema.ResourceData, meta interface{}) error {
+func resourceASConfigurationRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	asClient, err := config.AutoscalingV1Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
+		return diag.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_configuration.go]: pls remove huawei

File: huaweicloud/services/as/resource_huaweicloud_as_configuration.go [hunk 3]
```diff
@@ -301,51 +310,52 @@ func resourceASConfigurationCreate(d *schema.ResourceData, meta interface{}) err
...
 	return nil
 }
 
-func resourceASConfigurationDelete(d *schema.ResourceData, meta interface{}) error {
+func resourceASConfigurationDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	asClient, err := config.AutoscalingV1Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
+		return diag.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_configuration.go]: ditto

File: huaweicloud/services/as/resource_huaweicloud_as_lifecycle_hook.go [hunk 4]
```diff
@@ -89,12 +91,13 @@ func ResourceASLifecycleHook() *schema.Resource {
 	}
 }
 
-func resourceASLifecycleHookCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceASLifecycleHookCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	client, err := config.AutoscalingV1Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud AutoScaling client: %s", err)
+		return diag.Errorf("Error creating HuaweiCloud AutoScaling client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_lifecycle_hook.go]: pls remove huawei

File: huaweicloud/services/as/resource_huaweicloud_as_lifecycle_hook.go [hunk 5]
```diff
@@ -106,49 +109,51 @@ func resourceASLifecycleHookCreate(d *schema.ResourceData, meta interface{}) err
...
 }
 
-func resourceASLifecycleHookRead(d *schema.ResourceData, meta interface{}) error {
+func resourceASLifecycleHookRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	region := config.GetRegion(d)
 	client, err := config.AutoscalingV1Client(region)
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud AutoScaling client: %s", err)
+		return diag.Errorf("Error creating HuaweiCloud AutoScaling client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_lifecycle_hook.go]: ditto

File: huaweicloud/services/as/resource_huaweicloud_as_policy.go [hunk 6]
```diff
@@ -259,16 +273,16 @@ func resourceASPolicyRead(d *schema.ResourceData, meta interface{}) error {
 	return nil
 }
 
-func resourceASPolicyUpdate(d *schema.ResourceData, meta interface{}) error {
+func resourceASPolicyUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	asClient, err := config.AutoscalingV1Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
+		return diag.Errorf("Error creating HuaweiCloud autoscaling client: %s", err)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_policy.go]: ditto

File: huaweicloud/services/acceptance/as/resource_huaweicloud_as_configuration_test.go [hunk 7]
```diff
@@ -49,30 +45,28 @@ func testAccCheckASV1ConfigurationDestroy(s *terraform.State) error {
...
 		}
 	}
 
-	logp.Printf("[DEBUG] testAccCheckASV1ConfigurationDestroy success!")
-
 	return nil
 }
 
-func testAccCheckASV1ConfigurationExists(n string, configuration *configurations.Configuration) resource.TestCheckFunc {
+func testAccCheckASConfigurationExists(n string, configuration *configurations.Configuration) resource.TestCheckFunc {
```
Comment (hmbig2) [inline huaweicloud/services/acceptance/as/resource_huaweicloud_as_configuration_test.go:55]: pls use acceptance.InitResourceCheck instead

File: huaweicloud/services/acceptance/as/resource_huaweicloud_as_group_test.go [hunk 8]
```diff
@@ -96,30 +92,28 @@ func testAccCheckASV1GroupDestroy(s *terraform.State) error {
...
 		}
 	}
 
-	logp.Printf("[DEBUG] testCheckASV1GroupDestroy success!")
-
 	return nil
 }
 
-func testAccCheckASV1GroupExists(n string, group *groups.Group) resource.TestCheckFunc {
+func testAccCheckASGroupExists(n string, group *groups.Group) resource.TestCheckFunc {
```
Comment (hmbig2) [inline huaweicloud/services/acceptance/as/resource_huaweicloud_as_group_test.go:102]: pls use acceptance.InitResourceCheck instead

File: huaweicloud/services/as/resource_huaweicloud_as_configuration.go [hunk 9]
```diff
@@ -301,51 +310,52 @@ func resourceASConfigurationCreate(d *schema.ResourceData, meta interface{}) err
...
 	}
 
 	asConfig, err := configurations.Get(asClient, d.Id()).Extract()
 	if err != nil {
-		return common.CheckDeleted(d, err, "AS Configuration")
+		return common.CheckDeletedDiag(d, err, "AS Configuration")
 	}
 
 	logp.Printf("[DEBUG] Retrieved ASConfiguration %q: %+v", d.Id(), asConfig)
```
Comment (hmbig2) [inline huaweicloud/services/as/resource_huaweicloud_as_configuration.go:332]: s/logp.printf/log.printf

**Comment by ShiChangkuo** [comment]
@hmbig2 all of the `HuaweiCloud` in error messages are removed.

File: huaweicloud/services/as/resource_huaweicloud_as_configuration.go (hunk 9 above)
Comment (ShiChangkuo) [reply huaweicloud/services/as/resource_huaweicloud_as_configuration.go:332]: both logp and log are acceptable, keep consistent will be fine.

File: huaweicloud/services/acceptance/as/resource_huaweicloud_as_group_test.go (hunk 8 above)
Comment (ShiChangkuo) [reply huaweicloud/services/acceptance/as/resource_huaweicloud_as_group_test.go:102]: for now, we can keep it as is.

**Pull request merged**